src/parser/on_demand/parse_to_elements.ts

```diff
@@ -0,0 +1,131 @@
+import { BSONOffsetError } from '../../error';
+import { NumberUtils } from '../../utils/number_utils';
+
+/**
+ * @public
+ * @experimental
+ */
+export type BSONElement = [
+  type: number,
+  nameOffset: number,
+  nameLength: number,
+  offset: number,
+  length: number
+];
+
+/**
+ * Searches for null terminator.
+ * **Does not** bounds check since this should **ONLY** be used within parseToElements which has asserted that `bytes` ends with a `0x00`.
+ * So this will at most iterate to the document's terminator and error if that is the offset reached.
+ */
+function findNull(bytes: Uint8Array, offset: number): number {
+  let nullTerminatorOffset = offset;
+
+  for (; bytes[nullTerminatorOffset] !== 0x00; nullTerminatorOffset++);
+
+  if (nullTerminatorOffset === bytes.length - 1) {
+    throw new BSONOffsetError('Null terminator not found', offset);
+  }
+
+  return nullTerminatorOffset;
+}
+
+/**
+ * @public
+ * @experimental
+ */
+export function parseToElements(bytes: Uint8Array, startOffset = 0): Iterable<BSONElement> {
```

Comment:
Yes, this was in the design, we can in the future return a generator instead of an array. Probably based on the number of bytes or some other heuristic or explicit setting.